### ISSUE TITLE
New LoggerV2 interface: unclear who should call os.Exit for Fatal*

### ISSUE THREAD

**Issue by danjacques** [grpc/grpc-go#1360]
The [interface definition](https://github.com/grpc/grpc-go/blob/41d9b6ea2a6335f3a22074ed35c0542c9da1baf4/grpclog/loggerv2.go#L46) says that the `Fatal*` functions should call `os.Exit()`, e.g.:

```go
// Fatalf logs to ERROR log. Arguments are handled in the manner of fmt.Printf.
// This function should call os.Exit() with a non-zero exit code.
```

However, the [default implementation](https://github.com/grpc/grpc-go/blob/41d9b6ea2a6335f3a22074ed35c0542c9da1baf4/grpclog/loggerv2.go#L166) does not do this; instead, it is done at the [package interface level](https://github.com/grpc/grpc-go/blob/41d9b6ea2a6335f3a22074ed35c0542c9da1baf4/grpclog/grpclog.go#L90).

The `glog` implementation [calls](https://github.com/grpc/grpc-go/blob/41d9b6ea2a6335f3a22074ed35c0542c9da1baf4/grpclog/glogger/glogger.go#L70) `glog.Fatalf`, which fulfills this contract.

I think the default interface implementation needs to be updated to directly call `os.Exit` as per the interface.

**Comment by menghanl** [comment]
The implementation should call `os.Exit()`.

In the default implementation, `log.Logger.Fatalf` is called, which calls `os.Exit()` eventually.

The package interface level `os.Exit()` is called in case some implementation doesn't exit on fatal.
It is a no-op if the logging implementation does the right thing.

**Comment by danjacques** [comment]
It seems prudent to update the default implementation to conform to this then, right? ATM it does not call `os.Exit`.

**Comment by menghanl** [comment]
Calling `os.Exit()` again in the default implementation after `Fatal()` seems unnecessary to me.
Also the logger interface documentation makes it clear that implementation should call `os.Exit()`.

It seems the confusion is that, at the first glance, the default implementation doesn't call `os.Exit()`.
How about adding more comments in the default implementation and also in the package functions to make what happened on `os.Exit()` more clear? So that people won't get mislead by it?

**Comment by danjacques** [comment]
This is obviously not that big of a deal. The concern that I have is that the interface explicitly states:

```go
// This function should call os.Exit() with a non-zero exit code.
```

... and yet the canonical implementation of that interface doesn't do this. IMO, an appropriate solution is to update the interface comment to say something like, `This function may call os.Exit() to terminate. If it doesn't, os.Exit will be called after this function returns`. This makes it clear that the function is permitted to, but not required to, terminate the application.

**Comment by menghanl** [comment]
We do want the implementation to call os.Exit() in Fatal functions.
What do you think about the comments added in #1365?

**Comment by danjacques** [comment]
I think the comments in `grpclog/grpclog.go` helpfully document why there is an `os.Exit` as a failsafe.

However, I think the comments in `grpclog/loggerv2.go` don't resolve the underlying concern, which is that `grpclog.loggert` does not conform to the `grpclog.LoggerV2` interface's soft (comment) API. If the safety net is `gRPC` defending itself from improper `LoggerV2` implementations, then `grpclog.loggert` is improper and should be fixed. Otherwise, if it is permitted for a `LoggerV2` to rely on `gRPC`'s failsafe, the interface requirement should be softened from `should` to `may`.

Anyways, this is pretty trivial, and the comments in the patch do clarify that `os.Exit` is omitted as an optimization. Thanks for spending time on this.

For posterity, I encountered this problem b/c I have a custom `Logger` implementation that was written with the expectation that `os.Exit` was the full responsibility of the `Logger`. I relied on this behavior in a unit test, where I disabled the `Exit` call in my custom `Logger` so that I could assert that `Exit` was called without actually exiting my test. This started failing with the `LoggerV2` patch because of the introduction of the higher-level failsafe `os.Exit`, and when I dove into it, it was not clear to me whether my assumption was valid or not. As I mentioned above, I still don't think the responsibility is clear even with the additional comments, but they do shed some light on the design decision involved :)
